_gsocblogs/2022/blog_Acts_KavishkaAttanayake.md

```diff
@@ -0,0 +1,59 @@
+were discussed.
+
+#### Bi-weekly Acts Parallelization Group meeting
+
+Bi weekly group meetings where contributors from different organizations and institutes discuss their updates. Moreover, dicussions on any new projects are
+carried out. Understanding everything that is being discussed is a big challenge, nonetheless it is enjoyable listening to such an enthusiastic crowd.
+This is a valuable oppotunity for me to grasp whats going on in the community and also learn from their discussions.
+
+## Progress
+#### Weeks 1-4 ( including community bonding )
+> Ending on 16.06.2022 
+
+CUDA-MPS was something relatively new to me at the time, I had tested CUDA-MPS during application period and continued this work over during community bonding. Joined my first parallelization group meeting, gave me the opportunity to see and listen the developers and introduce myself as well. Meanwhile started working on porting SYCL clusterization code to CUDA and created a [PR (PR-206)]( https://github.com/acts-project/traccc/pull/206 ). This PR was not merged in as it was outdated at the time, nevertheless it was a great starting point and got valuable feedback from the community.
+Moreover, my mentors provided access to a server at Lawrence Berkeley National Laboratory to conduct the benchmarking.
+
+#### Weeks 5-8 
+> Ending on 14.07.2022
+
+Modified the PR-206 to fit in with the latest version and added suggested changes, this was merged in [PR-209]( https://github.com/acts-project/traccc/pull/209 ).
+Tests were done using caching allocators, and there were issues along the way ([issue with contiguous memory resource](https://github.com/acts-project/vecmem/issues/180) and issue with [binary page memory resource]( https://github.com/acts-project/vecmem/issues/182 )), Once the issues were fixed by the Vecmem developers. I carried out benchmarks with using contiguous memory resource and got expected results. Sadly, using binary page memory resource did worse than using any caching allocation at all. Therefore, benchmarking throughput with caching allocators was postponed. Meanwhile, a way to use Contiguous memory resource as an alternative will be explored. In addtion, prepared the bash scripts and modified the CUDA algorithm to be suitable for benchmarking and comparing against the CPU algorihtm.
+
+#### Weeks 9-10
+> Ending on 28.07.2022
+
+Benchmarked overall throughput for the CPU algorithm and CUDA algorithm with and without MPS [logs available here]( https://drive.google.com/drive/folders/15QFPNNwgh75RoRZ2au2_WybCuIMbUQpQ ). A significant improvement was not observable, since CUDA kernel execution only takes up ~10% of the entire process. As adviced by my mentors I started benchmarking kernel level throughput with and without MPS. Currently I have completed benchmarking and in the process of analyzing the kernel execution times.
```

Comment:
I think you can elaborate this section as it is a milestone for your project. Can you show some plots (throughtput vs. the number of processes) from the log and make some comments on why the improvement cannot be seen

> since CUDA kernel execution only takes up ~10% of the entire process.

How did you find this result?

Reply:
Sure i will add the logs as well.
The reason for the small increase in throughput is because kernel execution time is smaller than the wall time.   

I got that number from profiling and just the wall time of  `seq_example_cuda`. 
for instance,

```
build/bin/traccc_seq_example_cuda --detector_file=tml_detector/trackml-detector.csv --digitization_config_file=tml_detector/default-geometric-config-generic.json --cell_directory=tml_full/ttbar_mu200/ --events=1 --run_cpu=0 --input-binary
Running build/bin/traccc_seq_example_cuda tml_detector/trackml-detector.csv tml_full/ttbar_mu200/ 1
==> Statistics ... 
- read    92504 spacepoints from 13565 modules
- created        321488 cells           
- created        92504 meaurements     
- created        92504 spacepoints     
- created (cpu)  0 seeds
- created (cuda) 15424 seeds
==> Elpased time ... 
wall time           0.685223  
file reading (cpu)        0.0585111 
clusterization_time (cpu) 0.018244  
spacepoint_formation_time (cpu) 0.00126185
clusterization and sp formation (cuda) 0.0182342 
seeding_time (cpu)        4.081e-06 
seeding_time (cuda)       0.036016  
tr_par_esti_time (cpu)    1.55e-07  
tr_par_esti_time (cuda)   0.00196764

```
sum of cuda algorithms ~ 56ms (will be getting similar numbers under profiling as well)
wall time ~ 685ms
In this case its it spends about ~ 8% of its time doing kernel execution, and when we use MPS and multiple processe only this 8% of the time will be parallelized at the GPU. Therefore when comparing with and without MPS there is no major improvement in overall throughput.

"_A significant improvement was not observable_"
Sorry if my wording gave a wrong idea, I will re write it as "a significant improvement was not observable with using MPS compared to no MPS".